File: dashboard_backend/src/main/java/com/finance/dashboard/service/UserService.java
package com.finance.dashboard.service;

import com.finance.dashboard.exception.DuplicateResourceException;
import com.finance.dashboard.exception.ResourceNotFoundException;
import com.finance.dashboard.model.Role;
import com.finance.dashboard.model.User;
import com.finance.dashboard.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

@Service
public class UserService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private PasswordEncoder passwordEncoder;

    // Register new user
    public User createUser(User user) {
        if (userRepository.existsByEmail(user.getEmail())) {
            throw new DuplicateResourceException(
                    "Email already registered: " + user.getEmail());
        }
        user.setPassword(passwordEncoder.encode(user.getPassword()));
        user.setActive(true);
        if (user.getRole() == null) user.setRole(Role.VIEWER);
        return userRepository.save(user);
    }

    // Find by email
    public Optional<User> findByEmail(String email) {
        return userRepository.findByEmail(email);
    }

    // Get all users
    public List<User> getAllUsers() {
        return userRepository.findAll();
    }

    // Update role
    public User updateUserRole(Long userId, Role role) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "User not found with id: " + userId));
        user.setRole(role);
        return userRepository.save(user);
    }

    // Update status
    public User updateUserStatus(Long userId, Boolean active) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "User not found with id: " + userId));
        user.setActive(active);
        return userRepository.save(user);
    }

    // Delete user
    public void deleteUser(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException(
                        "User not found with id: " + userId));
        userRepository.delete(user);
    }
}

File: dashboard_frontend/src/services/userService.js
import API from "../api/api";

// Get all users
export const getUsers = async () => {
    const response = await API.get("/users");
    return response.data;
};

// Create user
export const createUser = async (userData) => {
    const response = await API.post(
        "/users",
        userData
    );

    return response.data;
};

// Update role
export const updateUserRole = async (id, role) => {
    const response = await API.put(
        `/users/${id}/role?role=${role}`
    );

    return response.data;
};

// Update active status
export const updateUserStatus = async (id, active) => {
    const response = await API.put(
        `/users/${id}/status?active=${active}`
    );

    return response.data;
};

// Delete user
export const deleteUser = async (id) => {
    const response = await API.delete(
        `/users/${id}`
    );

    return response.data;
};